### PR TITLE
chore(Data/Matrix/RowCol): use `Fin 1` instead of `Unit` in `Matrix.row` and `Matrix.col`

### PR DESCRIPTION
**Pull request by joneugster**
Use `Fin 1` instead of `Unit` in the definition of `Matrix.col` and `Matrix.row` in order to make it interact better with matrix notation `!![]`.

---

Currently, the following example does not type-check because it expects `Unit` but got `Fin 1`:

```lean
import Mathlib

example : Matrix.row ![1, 1] = !![1, 1] := by
  ext i j
  simp
```

This PR adds a new test case capturing the issue described above and fixes the new test by using `Fin 1` in said definitions.

See also #11536, which makes a more elaborate attempt to fix the same failing test case by generalising the definition to use an arbitrary `{ι : Type*}` instead of `Fin 1`.

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Why.20does.20.60Matrix.2Erow.60.20use.20.20.60Unit.60.20instead.20of.20.60Fin.201.60.3F)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Matrix/RowCol.lean [hunk 1]
```diff
@@ -27,7 +27,7 @@ variable {R : Type*} {α : Type v} {β : Type w}
 namespace Matrix
 
 /-- `Matrix.col u` is the column matrix whose entries are given by `u`. -/
-def col (w : m → α) : Matrix m Unit α :=
+def col (w : m → α) : Matrix m (Fin 1) α :=
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/RowCol.lean:30]: An alternative would be to have this defined for any `Unique ι`, which may or may not be annoying downstream.

Comment (YaelDillies) [reply Mathlib/Data/Matrix/RowCol.lean:30]: I was coming on this PR to say the same thing

Comment (joneugster) [reply Mathlib/Data/Matrix/RowCol.lean:30]: That sounds like the better idea indeed! Ill make these modifications tomorrow

Comment (joneugster) [reply Mathlib/Data/Matrix/RowCol.lean:30]: I tried using

```lean
def col {ι : Type} [Unique ι] (w : m → α) : Matrix m ι α :=
  of fun x _ => w x
```

but the implications were that any theorem about `col` and `row` anywhere needed addition of `{ι : Type} [Unique ι]` just to make the statement compile.

Here is a PR fixing just two files with the alternative choice, I think we would see a lot of `(ι := ι)` appear everywhere: #11536

Comment (joneugster) [reply Mathlib/Data/Matrix/RowCol.lean:30]: Update: #11536 is now also ready for review. The two PRs are mutually exclusive.

I have to admit, I still don't really understand what we gain by generalising which one-element-index-type we are using.

To me it seems that `Fin 1` is only correct choice: For concrete matrices we need `Fin 1` for the notation and if anybody ever defined an abstract matrix on a different type `ι` with `Unique ι`, they could very likely use `Equiv.equivPUnit` to turn that into `Fin 1` instead.

Comment (YaelDillies) [reply Mathlib/Data/Matrix/RowCol.lean:30]: Maybe you're right. It depends on how people use it

**Comment by joneugster** [comment]
!bench

**Comment by joneugster** [comment]
#11536 is now being merged, superseding this PR.

**Pull request closed without merging**